tests/integration-tests.test.ts

```diff
@@ -409,4 +409,109 @@ describe('integration tests', () => {
+             helloRouteHandler: SinonSpy,
+             helloRouteHandler2: SinonSpy,
+             goodbyeRouteHandler: SinonSpy,
+             helloReq: { url: string; path: string; query: KeyValueStringObject },
+             hello2Req: { url: string; path: string; query: KeyValueStringObject },
+             goodbyeReq: { url: string; path: string; query: KeyValueStringObject };
+
+         helloRouteHandler = spy((req: Request, _resp: Response, next: NextCallback): void => {
+            helloReq = _.pick(req, ...reqProps);
+            changeReqFn(req);
+            next();
+         });
+
+         helloRouteHandler2 = spy((req: Request, resp: Response): void => {
+            hello2Req = _.pick(req, ...reqProps);
+            resp.send('hello');
+         });
+
+         goodbyeRouteHandler = spy((req: Request, resp: Response): void => {
+            goodbyeReq = _.pick(req, ...reqProps);
+            resp.send('goodbye');
+         });
+
+         router.get('/hello', helloRouteHandler)
+            .get('/goodbye', goodbyeRouteHandler)
+            .get('/hello', helloRouteHandler2);
+
+         app.addSubRouter('/path', router);
+
+         testOutcome('GET', '/path/hello', expectedResponse);
+
+         return {
+            hello: helloRouteHandler,
+            getHelloReq: () => { return helloReq; },
+            hello2: helloRouteHandler2,
+            getHello2Req: () => { return hello2Req; },
+            goodbye: goodbyeRouteHandler,
+            getGoodbyeReq: () => { return goodbyeReq; },
+         };
+      }
+
+      it('redirects to a different route handler when req.url is changed', () => {
+         const handlers = testRoutesWithRedirect();
+
+         assert.calledOnce(handlers.hello);
+         assert.calledOnce(handlers.goodbye);
+         assert.notCalled(handlers.hello2);
+         assert.callOrder(handlers.hello, handlers.goodbye);
+      });
+
+      it('continues executing routes in the current chain before redirecting to the next match', () => {
+         const router = new Router();
+
+         let h1, h2Redirects, h3, h4ShouldBeSkipped, g1ShouldBeSkipped, g2Ends;
+
+         h1 = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+
+         h2Redirects = spy((req: Request, _resp: Response, next: NextCallback): void => {
+            req.url = '/goodbye';
+            next();
+         });
+
+         h3 = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+         h4ShouldBeSkipped = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+         g1ShouldBeSkipped = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+         g2Ends = spy((_req: Request, resp: Response): void => { resp.send('goodbye'); });
+
+         router.get('/goodbye', g1ShouldBeSkipped)
+            .get('/hello', h1, h2Redirects, h3)
+            .get('/hello', h4ShouldBeSkipped)
+            .get('/goodbye', g2Ends);
```

Comment:
In addition to the comment below (or instead of it), maybe a little commenting here would be helpful to the future reader (possibly your future self), e.g.

```
         router.get('/goodbye', g1ShouldBeSkipped) // because the request starts at /hello
            .get('/hello', h1, h2Redirects, h3) // even though h2 redirects to /goodbye, the chain should still run h3
            .get('/hello', h4ShouldBeSkipped) // by now the request is for /goodbye, so skip this
            .get('/goodbye', g2Ends); // this /goodbye gets called because of the redirect above
```